quinn-proto/src/endpoint.rs

```diff
@@ -560,7 +569,12 @@ where
                 return None;
             }
 
-            match RetryToken::from_bytes(&*server_config.token_key, &remote, &dst_cid, &token) {
+            let random_bytes = RetryToken::extract_random_bytes(&token);
+
+            // Derive AeadKey
+            let aead_key = S::AeadKey::from_hkdf(&*server_config.token_key, random_bytes);
+
+            match RetryToken::from_bytes(&aead_key, &remote, &dst_cid, &token) {
```

Comment:
And similarly for decoding. This should obviate the need to expose `extract_random_bytes`.